Many decoding fixes and added string formatting on PodInfoFaultEvent

Lytrix opened the pull request:
Updated Faultevents from latest message.md from itsmojo.
Fixed decoding errors, like the minutes -> days,  hours, minutes
Created string descriptions for enums to show readable values in the debugger.

ps2 commented on OmniKit/Model/FaultEventCode.swift:
```diff
@@ -17,12 +17,15 @@ public struct FaultEventCode: CustomStringConvertible {
         case failedFlashErase                     = 0x01
         case failedFlashStore                     = 0x02
         case tableCorruptionBasalExtraCommand     = 0x03
+        case faultEventType04                     = 0x04
```
FaultEventCode type already handles unknown types, that will stringify as "Fault Event Code %1$02x".  We shouldn't add any enums for types that we don't know anything about.

Lytrix replied:
Good idea, I was also a bit in doubt if this would be a good idea to add. I will remove the unknown ones.

ps2 commented on OmniKit/Model/FaultEventCode.swift:
```diff
@@ -120,194 +175,304 @@ public struct FaultEventCode: CustomStringConvertible {
             switch faultType {
             case .noFaults:
-                return LocalizedString("Fault Event Type 0x00: No faults, system running normally.", comment: "Fault Event Type 0x00: No faults, system running normally.")
+                return LocalizedString("Fault Event Type 0x00: No faults, system running normally", comment: "Fault Event Type 0x00: No faults, system running normally")
```
If you want to prefix *all* faults with "Fault Event Type 0xxx", you can split that out into a prefix generated by the FaultEventCode.description function.

ps2 commented on OmniKit/Model/FaultEventCode.swift:
```diff
@@ -120,194 +175,304 @@ public struct FaultEventCode: CustomStringConvertible {
+                return LocalizedString("Fault Event Type 0x00: No faults, system running normally", comment: "Fault Event Type 0x00: No faults, system running normally")
             case .failedFlashErase:
-                return LocalizedString("Fault Event Type 0x01: Flash erase failed in $4x00 page.", comment: "Fault Event Type 0x01: Flash erase failed in $4x00 page.")
+                return LocalizedString("Fault Event Type 0x01: Flash erase failed in $4x00 page", comment: "Fault Event Type 0x01: Flash erase failed in $4x00 page")
```
I'm not sure we want the Localize these strings.  Currently they don't surface in Loop, and I'm not sure they should.  Only if the user is going to do different things based on the error.  I think currently all these are debugging/diagnostic level.  I.e. they should be reported, and maybe included in issue reports, but at the top level, I think we tell the user "Pod internal fault xx.  Please generate issue report, deactivate and replace your pod".

Lytrix replied:
The issue reports will be the best place to use them eventually for sure, I will dive into the message types to figure out how to do this properly. Nice thing to look into next week.

Lytrix committed:
Removed periods + created prefix

Lytrix committed:
Renamed recieverLowGain->receiverLowGain

Lytrix committed:
Renamed recieverLowGain->receiverLowGain
Removed unused length enum

Lytrix committed:
- Fixed proper encoding of bits for radioRSSI and receiverLowGain
- Renamed receiverLowGain in IssueReport string

ps2 commented on OmniKit/Model/FaultEventCode.swift:
```diff
@@ -117,197 +117,205 @@ public struct FaultEventCode: CustomStringConvertible {
+                return "Fault Event Type 0x90: Bad value for $283/$2E3/$315 table specification"
             case .faultEventSetupPodType91:
-                return LocalizedString("Fault Event Type 0x91: Problem inside verify_and_start_pump.", comment: "Fault Event Type 0x91: Problem inside verify_and_start_pump.")
+                return "Problem inside verify_and_start_pump"
```
The string does not include the fault code, which is important.

ps2 commented on OmniKit/Model/FaultEventCode.swift:
```diff
@@ -117,197 +117,205 @@ public struct FaultEventCode: CustomStringConvertible {
             }
         } else {
-            return String(format: LocalizedString("Fault Event Code %1$02x", comment: "Format string for unknown Fault Event code."), rawValue)
+             return "Unknown Error Code"
```
This still needs to include the fault number

Lytrix replied:
Ah yes, that else statement is not in the variable function.

ps2 commented on OmniKit/Model/FaultEventCode.swift:
```diff
@@ -117,197 +117,205 @@ public struct FaultEventCode: CustomStringConvertible {
 }
+
+extension String {
+    mutating func addFaultCodePrefix(rawValue: UInt8) {
```
This is not being used.  Also, it can probably just be inline above, in the description var; no need to extend a system class here.

Lytrix committed:
Moved prefix function into var description.
Removed all hardcoded FaultEvent Type 0xXX from descriptions.

ps2 commented on OmniKit/Model/FaultEventCode.swift:
```diff
@@ -117,197 +117,199 @@ public struct FaultEventCode: CustomStringConvertible, Equatable {
             switch faultType {
             case .noFaults:
-                return LocalizedString("Fault Event Type 0x00: No faults, system running normally.", comment: "Fault Event Type 0x00: No faults, system running normally.")
+                return "\(prefix) No faults, system running normally"
```
```
let faultDescription: String
if let faultType = faultType {
    faultDescription = { 
        switch faultType {
        case .noFaults:
            return "No faults, system running normally"
        ....
        }
    }()
} else {
    faultDescription = "Unknown Error Code"
}
return String(format: "Fault Event Code %1$02x: %@", rawValue, faultDescription)
```

Lytrix replied:
I was also looking for a way to return a combined string below, but I did not know where to add the return string variable. So right before the switch, it seems so logical now but I could not find any examples of this. Thanks!

Lytrix committed:
Added string variable for switch result to return together with generated fault code.

ps2 commented on OmniKit/Model/FaultEventCode.swift:
```diff
@@ -117,197 +117,202 @@ public struct FaultEventCode: CustomStringConvertible, Equatable {
-            return String(format: LocalizedString("Fault Event Code %1$02x", comment: "Format string for unknown Fault Event code."), rawValue)
+             return "Fault Event Code %1$02x: Unknown Error Code"
         }
+        return String(format: "Fault Event Code %1$02x: %@", rawValue, faultDescription)
```
The 1$ can be removed

Lytrix committed:
Return Unknown Error as a faultDescription variable

ps2 commented on OmniKit/Model/FaultEventCode.swift:
```diff
@@ -311,7 +311,7 @@ public struct FaultEventCode: CustomStringConvertible, Equatable {
             }()
         } else {
-             return "Unknown Error Code"
+            faultDescription = "Fault Event Code %1$02x: Unknown Error Code"
```
This is not needed... The code is prepended below.

Lytrix replied:
I see it, I am committing too fast, time to go to bed. :-)

Lytrix committed:
Code cleanup and removed duplicate FaultError string on Unknown Error

Lytrix committed:
Be able to show negative RSSI and gain values by using an signed Integer.

ps2 commented on OmniKit/Model/FaultEventCode.swift:
```diff
@@ -117,197 +117,202 @@ public struct FaultEventCode: CustomStringConvertible, Equatable {
+            }()
         } else {
-            return String(format: LocalizedString("Fault Event Code %1$02x", comment: "Format string for unknown Fault Event code."), rawValue)
+            faultDescription = "Unknown Error Code"
```
"Unknown Error Code" -> "Unknown Fault"

Lytrix committed:
Renamed Unknown Error Code -> Unknown Fault

Lytrix committed:
Add missing rawType to testPumpAlarms assertion.

Lytrix committed:
Add missing rawType to pass testPumpAlarm

Lytrix committed:
Decode RSSI and Gain to Int8 values

Lytrix committed:
index on omnipod-testing-podinfo-updates: 63086ef Add missing rawType to testPumpAlarms assertion.

Lytrix committed:
- Renamed ReservoirStatus->PodProgressStatus
The name reservoir was confusing because it describes the progress from paring to deactivating of the Pod itself not just from the reservoir.

- Renamed originalLoggedFaultEvent->currentStatus
The naming is confusing because in reality it describes the current status where the pod is in.a

- Renamed secondaryLoggedFaultEvent->previousStatus
The naming was not a secondary log but the previous state the pod was in.

- Renamed insulineRemaining->ReservoirLevel
To be consistent with StatusResponse

- Reused the ReservoirLevel logic of decoding 51.15 to >50 from Statusresponse

- Renamed reservoirStatusAtFirstLoggedFaultEvent- > previousPodProgressStatus
To be consistent with using the PodProgressStatus

- Removed unused lengthType enum

- Decoded unknownPageCode to Hex value and fixed wrong decoding

- Let all PodInfoTests pass

Lytrix committed:
Renamed swift file ReseroirStatus->PodProgressStatus

Lytrix committed:
Merge branch 'omnipod-testing-fix-receiverLowGain-radioRSSI'

Conflicts:
	OmniKit/MessageTransport/MessageBlocks/PodInfoFaultEvent.swift
	OmniKitTests/PodInfoTests.swift

Lytrix committed:
Decoding fixed for timeActive.

Lytrix committed:
Fix failing Travis test: Renamed firstIndex -> index

Lytrix committed:
Removed 'rawType' in assertion, to pass Travis test.

Lytrix committed:
- Removed missed periods from faultevents
- Added >50 U to reservoirLevel debugDescription
- Removed unused function to parse time
- Added switch statement for day/days

Lytrix committed:
Updated PodInfotest to show String value for reservoirLevel

Lytrix committed:
- Added FaultEvent description
- Removed duplicate Status fields

Lytrix committed:
Added new Faulteventcodes.

Lytrix committed:
Merge branch 'omnipod-testing'

Lytrix committed:
Updated FaultEvents with latest message.md values

Lytrix committed:
Updated Loopkit in Cartfile.resolved to pass Travis build

Lytrix committed:
Added description variable for showing names of DeliveryTypes in debugger
instead of DeliveryType(rawValue:0)

Lytrix committed:
- Removed whitespaces
- Added enum for logEventErrorType to switch the different ErrorTypes instead of separate true/false evaluations.
Still need to test if multiple ErrorTypes can be returned, no it assumes there can only be one.
- Fixed missing last to bytes (YYYY) in encodedData PodInfoResponse
- Added missing YYYY unknown Value
- Renamed previousPodProgressStatus -> logEventErrorPodProgressStatus
to let the PodProgressStatus be a part of the information about logged fault event
- Renamed previousPodProgressStatusCheck -> previousPodProgressStatus

Lytrix committed:
- Fixed seconds to minutes baseline for TimeInterval parsing
- Parse minutes correctly to day/hour/minute integers
- Renamed 'time since activation' strings to: 1 day plus 18:39
- Added tests to check proper string formatting for time debugger strings

Lytrix committed:
Added missing 'plus' on time sting test

ps2 commented on MinimedKit/Messages/GetPumpFirmwareVersionMessageBody.swift:
```diff
@@ -12,7 +12,7 @@ public class GetPumpFirmwareVersionMessageBody: CarelinkLongMessageBody {
     public let version: String
     
     public required init?(rxData: Data) {
-        let stringEnd = rxData.firstIndex(of: 0) ?? rxData.count
```
What is this change for?  I don't see an index(of:) function on Data in the docs.

Lytrix replied:
I had a Travis build error otherwise, it could not find firstIndex and suggested replacing it by index to pass the test. But running it locally it ran ok with firstIndex.

ps2 replied:
Ok, that was probably travis using Xcode 9.  You can change this back; we're running Xcode 10.

ps2 commented on OmniKit/MessageTransport/MessageBlocks/PodInfoFaultEvent.swift:
```diff
@@ -10,118 +10,133 @@ import Foundation
+        let resHighBits = Int(encodedData[11] & 0x03) << 6
+        let resLowBits = Int(encodedData[12] >> 2)
+        let reservoirValue = round(Double((resHighBits + resLowBits) * 50)/255)
+        if reservoirValue < StatusResponse.maximumReservoirReading {
```
This should be set to nil, if we are not at the point where we have a valid reservoir reading.  This is already within the context of having data, so we do not need to distinguish between not being able to get status and this field being set to >50.

Lytrix replied:
Ok, changed this be set to nil.

ps2 commented on OmniKit/MessageTransport/MessageBlocks/PodInfoFaultEvent.swift:
```diff
@@ -10,118 +10,133 @@ import Foundation
         
-        self.insulinRemaining = podPulseSize * Double((Int(encodedData[11] & 0x3) << 8) | Int(encodedData[12]))
+        let resHighBits = Int(encodedData[11] & 0x03) << 6
+        let resLowBits = Int(encodedData[12] >> 2)
```
Did we determine that these 2 bits do not have valid info?  I meant to look into this.

Lytrix replied:
These 2 bits do return the insulinRemaining. I've tested this with a Pod failure where there was 0.05 units not delivered:
`0216020f0000010200ec6a026803ff026b000028a70820230169` 
I will add this as a Unit test.

ps2 commented on OmniKit/MessageTransport/MessageBlocks/PodInfoFaultEvent.swift:
```diff
@@ -10,118 +10,133 @@ import Foundation
-        guard let reservoirStatusAtFirstLoggedFaultEventType = ReservoirStatus(rawValue: encodedData[17] & 0xF) else {
-            throw MessageError.unknownValue(value: encodedData[17] & 0xF, typeDescription: "ReservoirStatus")
+        guard let logEventErrorType = LogEventErrorType(rawValue:encodedData[17] >> 4) else {
+            throw MessageError.unknownValue(value: encodedData[17] >> 4, typeDescription: "LogEventErrorType")
```
I think this should maybe not throw an exception. I'm guessing we are likely to encounter new values for this, so we should just record it as an unknown type, like we do for FaultCodes.  If we're very confident of not receiving new, unknown values here, then we can throw.

ps2 commented on RileyLink.xcodeproj/project.pbxproj:
```diff
@@ -2863,6 +2863,7 @@
 					};
 					431CE7761F98564200255374 = {
 						CreatedOnToolsVersion = 9.0;
+						DevelopmentTeam = 86HTWX383F;
```
Change these back to unset.

Lytrix committed:
- Return nil for ReservoirLevel if it is not <50 U
- Ensure new unknown logEventError values are printed properly.
Setup logEventErrorType the same as faultEventType as a Model.
This way it will return a value as a description or as an unknown code value.

Lytrix committed:
Renamed index -> firstIndex back as it is setup in Data.
Travis build error solved by updating from v9 to v10.

Lytrix committed:
Unset all Development_team values in project.

Lytrix committed:
Added Unit Test PodInfoFaultEvent: InsulinNotDelivered

Lytrix committed:
Merge branch 'omnipod-testing' into omnipod-testing-podinfo-updates

itsmojo commented:
To confirm, the entire 10 reservoir field is either $3ff (> 50U) or the number of pulses left (which should only be 0-1000 decimal). I'm not a Swift guy, but I'm not sure why the code is so messy with multiple computations and the rounding. Just extract the 10 bits once into a local variable and if it's not 0x3ff, then set the insulin remaining to be value*podPulseSize. The $3ff is a valid value indicating >= 50 U. Perhaps we might want to differentiate getting the valid > 50 U value and not getting any valid status, but I don't have a strong opinion right now as I'm not familiar with the Swift code implementations.

ps2 commented:
Yes, the comment about looking into those lower bits was that this code had been written before the field was described as well as it is now in the wiki.  Sounds like we should do exactly what you say in terms of extracting the 10 bits as # of pulses, and multiplying by podPulseSize.  As for not getting any valid status, in this context we certainly do have a valid status, so we only need to distinguish between valid data (<= 50) and > 50 (no idea where above 50).

Lytrix committed:
Simplified the reservoirValue to read the decimal * podPulseSize
And showing the value 50 as well.

Lytrix commented:
@itsmojo and @ps2 
Thanks, for the input.
I've cleaned up the code as you stated, but kept in the validation, so anything above 1000 returns nil in this case. I did add a <= so it can start showing 50 if you say that 1000 decimal is also possible.
I will also change it on the StatusResponse part where I got this part of the code from, if this looks ok.

Lytrix committed:
- Fixed decoding only first 10 bits to count reservoirValue
- Added StatusResponse test with 46U left. (real data capture)

ps2 commented on OmniKit/MessageTransport/MessageBlocks/PodInfoFaultEvent.swift:
```diff
@@ -10,118 +10,106 @@ import Foundation
-        self.reservoirStatusAtFirstLoggedFaultEventCheck = reservoirStatusAtFirstLoggedFaultEventCheckType
+        self.previousPodProgressStatus = previousPodProgressStatus
+        
+        self.unKnownValue = encodedData[20...21]
```
unKnownValue -> unknownValue

ps2 commented on OmniKit/MessageTransport/MessageBlocks/PodInfoFaultEvent.swift:
```diff
@@ -10,118 +10,106 @@ import Foundation
-            "immediateBolusInProgress: \(immediateBolusInProgress)",
+            "recieverLowGain: \(receiverLowGain)",
+            "previousPodProgressStatus: \(previousPodProgressStatus)",
+            "unKnownValue: \(unKnownValue)",
```
This is a Data object right? Do you want `unKnownValue.hexadecimalString`?

Lytrix replied:
Ah yes, that's what's needed for now.

ps2 commented on OmniKit/Model/PodProgressStatus.swift:
```diff
@@ -1,5 +1,5 @@
 //
-//  ReservoirStatus.swift
+//  PodState.swift
```
PodProgressStatus

ps2 commented on RileyLink.xcodeproj/project.pbxproj:
```diff
@@ -2926,7 +2929,7 @@
 					C1FFAF7F213323CC00C50C1D = {
 						CreatedOnToolsVersion = 9.4.1;
-						DevelopmentTeam = UY678SP37Q;
+						DevelopmentTeam = 86HTWX383F;
```
Seems like there are still a few of these?

Lytrix replied:
I removed them from the UI but apparently there was still one left. Thanks for catching that.

ps2 replied:
There are a couple more at the bottom of the diff.

Lytrix committed:
Merge remote-tracking branch 'upstream/omnipod-testing' into omnipod-testing-podinfo-updates

Lytrix committed:
Fixed typo unKnownValue->unknownValue

Lytrix committed:
- Added hexadecimalString formatting to unknownvalue
- Updated filename title of swift file to PodProgressStatus.swfit
- Removed last development team value from project

Lytrix committed:
Updated Cartfile.resolved to latest LoopKit to pass Travis
